File: agent/nodes/assignTickets.js
const { assignAndStart } = require('../tools/jira');

/**
 * ASSIGN TICKETS node — assign user1 → first ticket, user2 → second ticket.
 * Picks the first unprocessed ticket as the currentTicket for this run.
 * @param {import('../state').AgentStateAnnotation} state
 */
async function assignTicketsNode(state) {
  console.log('\n👥 NODE: assignTickets — assigning tickets to users...');

  const { tickets } = state;
  if (!tickets || tickets.length === 0) {
    console.log('  ⚠️  No tickets to assign.');
    return { ...state, currentTicket: null };
  }

  const user1Id = process.env.JIRA_USER1_ACCOUNT_ID;
  const user2Id = process.env.JIRA_USER2_ACCOUNT_ID;

  const assignedTickets = {};

  // Assign ticket 0 to user1, ticket 1 to user2
  const pairs = [
    { ticket: tickets[0], userId: user1Id, userKey: 'user1' },
    { ticket: tickets[1], userId: user2Id, userKey: 'user2' },
  ].filter((p) => p.ticket && p.userId);

  for (const { ticket, userId, userKey } of pairs) {
    try {
      await assignAndStart(ticket.key, userId);
      assignedTickets[userKey] = ticket;
      console.log(`  ✅ ${ticket.key} assigned to ${userKey}`);
    } catch (err) {
      console.error(`  ❌ Failed to assign ${ticket.key}:`, err.message);
      assignedTickets[userKey] = ticket;
    }
  }

  // Process user1's ticket first
  const currentTicket = assignedTickets.user1 ?? assignedTickets.user2 ?? null;
  console.log(`  🎯 Processing ticket: ${currentTicket?.key}`);

  return { ...state, assignedTickets, currentTicket };
}

module.exports = { assignTicketsNode };

File: python-backend/nodes/assign_tickets.py
import os
import base64
import httpx
from state import AgentState
from tools.logger import jira_log

USERS = ["user1", "user2", "user3"]


def _jira_headers() -> dict:
    email = os.environ["JIRA_EMAIL"]
    token = os.environ["JIRA_API_TOKEN"]
    encoded = base64.b64encode(f"{email}:{token}".encode()).decode()
    return {
        "Authorization": f"Basic {encoded}",
        "Content-Type": "application/json",
        "Accept": "application/json",
    }


async def _move_to_active_sprint(issue_key: str) -> None:
    """Add ticket to the active sprint so it shows on the Jira board."""
    base_url = os.environ.get("JIRA_BASE_URL", "")
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            r = await client.get(
                f"{base_url}/rest/agile/1.0/board/1/sprint",
                headers=_jira_headers(),
                params={"state": "active"},
            )
            r.raise_for_status()
            sprints = r.json().get("values", [])
            if not sprints:
                return
            sprint_id = sprints[0]["id"]
            await client.post(
                f"{base_url}/rest/agile/1.0/sprint/{sprint_id}/issue",
                headers=_jira_headers(),
                json={"issues": [issue_key]},
            )
            print(f"  🏃 Added {issue_key} to sprint {sprint_id}")
    except Exception as e:
        print(f"  ⚠️  Sprint assignment failed (non-fatal): {e}")


async def assign_tickets_node(state: AgentState) -> AgentState:
    print("\n👥 NODE: assignTickets — assigning tickets to users...")
    tickets = state.get("tickets") or []

    if not tickets:
        print("  ⚠️  No tickets to assign.")
        return {**state, "assigned_tickets": {}, "current_ticket": None}

    assigned: dict = {}
    for i, ticket in enumerate(tickets):
        user = USERS[i % len(USERS)]
        assigned[ticket["key"]] = user

    current = tickets[0]
    key = current["key"]
    print(f"  ✅ {key} assigned to {assigned[key]}")
    print(f"  🎯 Processing ticket: {key}")

    await _move_to_active_sprint(key)
    await jira_log(key, f"🚀 Pipeline Started — ticket {key} picked up by SMS Agent")

    return {**state, "assigned_tickets": assigned, "current_ticket": current}
